io-hdfs/src/main/java/cz/o2/proxima/storage/hdfs/HdfsBatchLogObservable.java

```diff
@@ -0,0 +1,417 @@
+    return new Path(
+        new URI(getURI().toString() + DIR_FORMAT.format(LocalDateTime.ofInstant(
+            d, ZoneId.ofOffset("UTC", ZoneOffset.ofHours(0))))
+            + "/" + toFinalName(minStamp, maxStamp)));
+  }
+
+  @VisibleForTesting
+  Path toTmpLocation(long part)
+      throws UnknownHostException, URISyntaxException {
+
+    // place the final file in directory /.tmp/
+    return new Path(
+        new URI(getURI().toString() + "/.tmp/"
+            + toPartName(part)));
+  }
+
+  private <T> T getCfg(
+      String name, Map<String, Object> cfg, Function<Object, T> convert, T defVal) {
+
+    return Optional.ofNullable(cfg.get(name)).map(convert).orElse(defVal);
+  }
+
+  private void clearTmpDir() {
+    try {
+      Path tmpDir = new Path(getURI().toString() + "/.tmp/");
+      if (getFs().exists(tmpDir)) {
+        RemoteIterator<LocatedFileStatus> files = getFs().listFiles(tmpDir, false);
+        String localhost = getLocalhost();
+        while (files.hasNext()) {
+          LocatedFileStatus file = files.next();
+          if (matchesHostname(localhost, file)) {
+            getFs().delete(file.getPath(), false);
+          }
+        }
+      }
+    } catch (IOException ex) {
+      log.warn("Failed to clean tmp dir", ex);
+    }
+  }
+
+  private boolean matchesHostname(String hostname, LocatedFileStatus file) {
+    return file.getPath().getName().endsWith("-" + hostname);
+  }
+
+  private String getLocalhost() throws UnknownHostException {
+    return InetAddress.getLocalHost().getCanonicalHostName();
+  }
+
+  @Override
+  public List<Partition> getPartitions(long startStamp, long endStamp) {
+    List<Partition> partitions = new ArrayList<>();
+    try {
+      RemoteIterator<LocatedFileStatus> it;
+      it = getFs().listFiles(new Path(getURI().toString()), true);
+      HdfsPartition current = null;
+      while (it.hasNext()) {
+        LocatedFileStatus file = it.next();
+        if (file.isFile()) {
+          if (current == null) {
+            current = new HdfsPartition(partitions.size());
+          }
+          Map.Entry<Long, Long> minMaxStamp = getMinMaxStamp(file.getPath().getName());
+          long min = minMaxStamp.getKey();
+          long max = minMaxStamp.getValue();
+          if (max >= startStamp && min <= endStamp) {
+            current.add(file);
+            if (current.size() > batchProcessSize) {
+              partitions.add(current);
+              current = null;
+            }
+          }
+        }
+      }
+      if (current != null && current.size() > 0) {
+        partitions.add(current);
+      }
+    } catch (IOException ex) {
+      throw new RuntimeException(ex);
+    }
+    return partitions;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void observe(
+      List<Partition> partitions,
+      List<AttributeDescriptor<?>> attributes,
+      BatchLogObserver observer) {
+
+    context.getExecutorService().execute(() -> {
```

Comment:
This might create new executor service each time `observe` is called. A more suitable approach would be to cache the returned value in a `transient` field.

Reply:
Based on what is happening in `io-hbase`, I thought it is desired to possibly give each request for observable separate resources. Depending on executor factory.

What is the contract then? Must be single executor, cached and thread safe for lazy init? It is not obvious to me from interfaces.

Reply:
Yep, `HBaseLogObservable` is exactly correct. The executor is cached in transient field. The contract is not to create the executor for each observer, but to reuse the executor for multiple tasks. Don't know how to make this obvious from interfaces, though.

Reply:
Done